lucene/misc/src/test/org/apache/lucene/misc/store/TestDirectIODirectory.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.lucene.misc.store;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import org.apache.lucene.document.*;
+import org.apache.lucene.index.DirectoryReader;
+import org.apache.lucene.index.IndexReader;
+import org.apache.lucene.index.RandomIndexWriter;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.PhraseQuery;
+import org.apache.lucene.store.*;
+import org.junit.BeforeClass;
+
+public class TestDirectIODirectory extends BaseDirectoryTestCase {
+
+  @BeforeClass
+  public static void checkSupported() {
+    assumeTrue(
+        "This test required a JDK version that has support for ExtendedOpenOption.DIRECT",
+        DirectIODirectory.ExtendedOpenOption_DIRECT != null);
+  }
+
+  @Override
+  protected DirectIODirectory getDirectory(Path path) throws IOException {
+    return new DirectIODirectory(
+        FSDirectory.open(path), DirectIODirectory.DEFAULT_MERGE_BUFFER_SIZE, 0L) {
+      @Override
+      protected boolean useDirectIO(IOContext context) {
+        return true;
+      }
+    };
+  }
+
+  public void testIndexWriteRead() throws IOException {
+    try (Directory dir = getDirectory(createTempDir("testDirectIODirectory"))) {
```

Comment:
It is very good that this test uses the "hack" DirectIODirectory that ALWAYS uses our implementation! By default it would only use it for merges, but here we enforce use of IndexInput/IndexOutput for all transactions (although they may be slower).

Reply:
Yup agreed! Otherwise subtle bugs like above could easily go undetected for a long time and hard to debug later.